distributed/worker_memory.py

```diff
@@ -234,25 +234,47 @@ def _maybe_pause_or_unpause(self, worker: Worker, memory: int) -> None:
             worker.status = Status.running
 
     async def _maybe_spill(self, worker: Worker, memory: int) -> None:
+        """If process memory is above the ``spill`` threshold, evict keys until it goes
+        below the ``target`` threshold
+        """
         if self.memory_spill_fraction is False:
             return
 
         # SpillBuffer or a duct-type compatible MutableMapping which offers the
         # fast property and evict() methods. Dask-CUDA uses this.
         if not hasattr(self.data, "fast") or not hasattr(self.data, "evict"):
             return
-        data = cast(ManualEvictProto, self.data)
 
         assert self.memory_limit
         frac = memory / self.memory_limit
         if frac <= self.memory_spill_fraction:
             return
 
-        total_spilled = 0
         worker_logger.debug(
             "Worker is at %.0f%% memory usage. Start spilling data to disk.",
             frac * 100,
         )
+
+        def metrics_callback(label: Hashable, value: float, unit: str) -> None:
+            if not isinstance(label, tuple):
+                label = (label,)
+            worker.digest_metric(("memory-monitor", *label, unit), value)
+
+        # Work around bug with Tornado PeriodicCallback, which does not properly
+        # insulate the ContextVar used by context_manager
```

Comment:
I can reproduce this reliably with the dask-xgboost-nyctaxi Feature Engineering notebook - you can see metrics coming from Worker.execute tagged with memory-monitor, which should be spill-only.
I failed to reproduce the issue in a unit test.